distributed/scheduler.py

```diff
@@ -5636,3 +5659,21 @@ def remove_worker(self, worker=None, **kwargs):
 
     def teardown(self):
         self.bcomm.close()
+
+
+class CollectTaskMetaDataPlugin(SchedulerPlugin):
+    def __init__(self, scheduler, name):
+        self.scheduler = scheduler
+        self.name = name
+        self.keys = set()
+        self.metadata = {}
+
+    def update_graph(self, scheduler, dsk=None, keys=None, restrictions=None, **kwargs):
+        self.keys.update(keys)
+
+    def transition(self, key, start, finish, *args, **kwargs):
+        if start == "processing" and (finish == "memory" or finish == "erred"):
```

Comment:
Consider also `long-running`.  I think that it might make sense to do this for all tasks that end up in a finished state, regardless of their start state.

Reply:
IIUC long running tasks on a worker are kept in the `processing` state on the scheduler, so this should include those tasks. But I agree with your larger point, this should handle all finished tasks regardless of their start state

Reply:
Ah!  I just checked and you're right.  My information is out-dated here.